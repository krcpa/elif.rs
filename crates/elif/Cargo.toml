[package]
name = "elif"
version = "0.8.2"
edition = "2021"
description = "Modern Rust web framework designed for exceptional developer experience"
license = "MIT"
authors = ["krcpa <krcpa@users.noreply.github.com>"]
repository = "https://github.com/krcpa/elif.rs"
homepage = "https://github.com/krcpa/elif.rs"
documentation = "https://docs.rs/elif"
keywords = ["web", "framework", "http", "async", "api"]
categories = ["web-programming::http-server"]

[dependencies]
# Re-export sub-packages (umbrella package depends on all others)
elif-core = { version = "0.6.1", path = "../core" }
elif-http = { version = "0.8.2", path = "../elif-http" }
elif-orm = { version = "0.7.0", path = "../orm" }
elif-auth = { version = "0.4.0", path = "../elif-auth" }
elif-cache = { version = "0.3.0", path = "../elif-cache" }
elif-macros = { version = "0.1.1", path = "../elif-macros" }

# These are no longer needed since macros are in separate crate
# proc-macro2 = "1.0"
# quote = "1.0"
# syn = { version = "2.0", features = ["full"] }

# Common dependencies for prelude
serde = { workspace = true }
serde_json = { workspace = true }
tokio = { workspace = true }
async-trait = { workspace = true }
env_logger = "0.10"

# Remove proc-macro from main crate - we'll create a separate macro crate later
# [lib]
# proc-macro = true

[dev-dependencies]
tokio-test = "0.4"