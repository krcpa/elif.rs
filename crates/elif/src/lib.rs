//! # elif.rs - The Rust web framework
//!
//! elif.rs is a modern web framework designed for both exceptional developer
//! experience and AI-native development.
//!
//! This is the main umbrella package that provides a unified API and convenient
//! imports for the entire elif.rs ecosystem.

// Re-export all sub-packages as modules
pub use elif_auth as auth;
pub use elif_cache as cache;
pub use elif_core as core;
pub use elif_http as http;
pub use elif_orm as orm;

// Re-export common types at root level for convenience
pub use elif_http::request::ElifRequest as Request;
pub use elif_http::response::ElifResponse as Response;
pub use elif_http::routing::ElifRouter as Router;
pub use elif_http::Server;
pub use elif_http::{HttpError, HttpResult};

// Re-export core functionality
pub use elif_core::{
    ApiError, ApiErrorResponse, AppConfig, AppConfigTrait, BaseModule, ConfigSource, Container,
    ContainerBuilder, CoreError, Environment, ErrorDefinition, Module, ModuleLoader,
    ModuleRegistry, ProviderRegistry, ServiceProvider, ServiceRegistry, ServiceScope,
};

// Prelude module for convenient imports
pub mod prelude;

// Macro functionality - re-export from elif-macros crate
pub use elif_macros as macros;
pub use elif_macros::main;

/// Current version of elif.rs
pub const VERSION: &str = env!("CARGO_PKG_VERSION");

/// Framework information
pub const FRAMEWORK_NAME: &str = "elif.rs";

/// Get framework version
pub fn version() -> &'static str {
    VERSION
}

/// Get framework name
pub fn name() -> &'static str {
    FRAMEWORK_NAME
}
