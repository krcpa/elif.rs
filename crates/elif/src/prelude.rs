//! # Prelude
//!
//! The prelude module provides convenient imports for common elif.rs functionality.
//!
//! ```rust
//! use elif::prelude::*;
//! ```

// Essential HTTP types
pub use crate::{HttpError, HttpResult};
pub use crate::{Router, Server};
pub use elif_http::{ElifRequest as Request, ElifResponse as Response};
// Also export original types for macro compatibility
pub use elif_http::{ElifRequest, ElifResponse};

// Common traits - using correct exports
pub use elif_http::GenericHandler as Handler;
pub use elif_http::{IntoElifResponse, Middleware};
pub use elif_orm::Model;

// Core types
pub use crate::{
    ApiError, ApiErrorResponse, AppConfig, AppConfigTrait, BaseModule, ConfigSource, Container,
    ContainerBuilder, CoreError, Environment, ErrorDefinition, Module, ModuleLoader,
    ModuleRegistry, ProviderRegistry, ServiceProvider, ServiceRegistry, ServiceScope,
};

// Utility functions - check if these exist or create aliases
// pub use elif_http::response::{response};
// pub use elif_http::request::{request};

// JSON helper
pub use serde_json::json;

// Common derives
pub use serde::{Deserialize, Serialize};

// Async traits
pub use async_trait::async_trait;

// Macros
pub use crate::main;
pub use elif_macros::bootstrap;
pub use crate::{controller, module};
pub use elif_http_derive::{get, post, put, delete, patch, head, options, param, body};

// Bootstrap traits and types
pub use elif_http::AppBootstrap;
pub use elif_http::bootstrap::AppBootstrapper;
