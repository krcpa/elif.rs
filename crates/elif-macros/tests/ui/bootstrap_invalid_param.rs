use elif::prelude::*;

#[elif::bootstrap(AppModule, invalid_param = "test")]
async fn main() -> Result<(), HttpError> {
    // This should fail because invalid_param is not a recognized parameter
}