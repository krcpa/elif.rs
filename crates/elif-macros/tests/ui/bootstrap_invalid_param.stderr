error: Unknown bootstrap parameter 'invalid_param'. Valid parameters are: addr, config, middleware

       ðŸ’¡ Usage examples:
       â€¢ #[elif::bootstrap] (auto-discovery)
       â€¢ #[elif::bootstrap(addr = "127.0.0.1:3000")]
       â€¢ #[elif::bootstrap(config = my_config())]
       â€¢ #[elif::bootstrap(middleware = [cors(), auth()])]
       â€¢ #[elif::bootstrap(AppModule)] (backward compatibility)
 --> tests/ui/bootstrap_invalid_param.rs:3:30
  |
3 | #[elif::bootstrap(AppModule, invalid_param = "test")]
  |                              ^^^^^^^^^^^^^

warning: unused import: `elif::prelude::*`
 --> tests/ui/bootstrap_invalid_param.rs:1:5
  |
1 | use elif::prelude::*;
  |     ^^^^^^^^^^^^^^^^
  |
  = note: `#[warn(unused_imports)]` on by default

error[E0601]: `main` function not found in crate `$CRATE`
 --> tests/ui/bootstrap_invalid_param.rs:6:2
  |
6 | }
  |  ^ consider adding a `main` function to `$DIR/tests/ui/bootstrap_invalid_param.rs`
